src/components/diagrams/diagram-common.js

```diff
@@ -0,0 +1,274 @@
+            display: 'block',
+            width: '100%',
+            height: '100%',
+        },
+        '& .nad-label-box': {
+            color: theme.palette.text.primary,
+            'font-family': theme.typography.fontFamily,
+        },
+        '& .nad-text-edges': {
+            stroke: theme.palette.text.primary,
+        },
+        overflow: 'hidden',
+        ...customStyle,
+    };
+};
+
+export const commonDiagramStyle = (theme, customStyle) => {
+    return {
+        divInvalid: {
+            '& .sld-active-power, .sld-reactive-power, .sld-voltage, .sld-angle':
+                {
+                    opacity: INVALID_LOADFLOW_OPACITY,
+                },
+            '& .nad-edge-infos': {
+                opacity: NAD_INVALID_LOADFLOW_OPACITY,
+            },
+        },
+        paperBorders: {
+            borderLeft: '1px solid ' + theme.palette.action.disabled,
+            borderBottom:
+                '1px solid ' + theme.palette.action.disabledBackground,
+            borderRight: '1px solid ' + theme.palette.action.hover,
+        },
+        ...customStyle,
+    };
+};
+
+export const ViewState = {
+    PINNED: 'pinned',
+    MINIMIZED: 'minimized',
+    OPENED: 'opened',
+};
+
+export const SubstationLayout = {
+    HORIZONTAL: 'horizontal',
+    VERTICAL: 'vertical',
+    SMART: 'smart',
+    SMARTHORIZONTALCOMPACTION: 'smartHorizontalCompaction',
+    SMARTVERTICALCOMPACTION: 'smartVerticalCompaction',
+};
+
+export const SvgType = {
+    VOLTAGE_LEVEL: 'voltage-level',
+    SUBSTATION: 'substation',
+    NETWORK_AREA_DIAGRAM: 'network-area-diagram',
+};
+
+export function getEquipmentTypeFromFeederType(feederType) {
+    switch (feederType) {
+        case 'LINE':
+            return equipments.lines;
+        case 'LOAD':
+            return equipments.loads;
+        case 'BATTERY':
+            return equipments.batteries;
+        case 'DANGLING_LINE':
+            return equipments.danglingLines;
+        case 'GENERATOR':
+            return equipments.generators;
+        case 'VSC_CONVERTER_STATION':
+            return equipments.vscConverterStations;
+        case 'LCC_CONVERTER_STATION':
+            return equipments.lccConverterStations;
+        case 'HVDC_LINE':
+            return equipments.hvdcLines;
+        case 'CAPACITOR':
+        case 'INDUCTOR':
+            return equipments.shuntCompensators;
+        case 'STATIC_VAR_COMPENSATOR':
+            return equipments.staticVarCompensators;
+        case 'TWO_WINDINGS_TRANSFORMER':
+        case 'TWO_WINDINGS_TRANSFORMER_LEG':
+        case 'PHASE_SHIFT_TRANSFORMER':
+            return equipments.twoWindingsTransformers;
+        case 'THREE_WINDINGS_TRANSFORMER':
+        case 'THREE_WINDINGS_TRANSFORMER_LEG':
+            return equipments.threeWindingsTransformers;
+        default: {
+            console.log('bad feeder type ', feederType);
+            return null;
+        }
+    }
+}
+
+export const useDiagram = () => {
```

Comment:
i suggest to optimise useDiagram by making some changes : 

- Combine similar useCallback functions: The openDiagramView, togglePinDiagramView, minimizeDiagramView, and closeDiagramView functions all dispatch the same type of action, but with different parameters. These functions can be combined into a single function that takes an additional parameter for the type of action to dispatch (open, togglePin, minimize, or close).

- Use the useEffect hook with dependencies: Currently, the useEffect hook is listening to changes in the diagramStates and studyUuid, but these values are not used inside the callback function. 
Instead, you can use the useEffect hook with a dependency array to only call the syncDiagramStateWithSessionStorage function when the specific action that causes the state to change is dispatched.

- Use useMemo instead of useCallback: Since all the functions are only calling dispatch with the same set of dependencies, useMemo is a more appropriate hook.

- Use a switch case inside the callback function of useEffect to call the appropriate function based on the action type

- consider using reselect to avoid recomputing the same thing everytime

- consider using Immer to update state immutablely

here's an example how this function can be:

```
export const useDiagram = () => {
    const dispatch = useDispatch();
    const {diagramStates, studyUuid, action} = useSelector((state) => ({
        diagramStates: state.diagramStates,
        studyUuid: state.studyUuid,
        action: state.action
    }));

    useEffect(() => {
        if(action === 'SYNC_DIAGRAM_STATE'){
           syncDiagramStateWithSessionStorage(diagramStates, studyUuid);
        }
    }, [action]);

    const handleDiagramView = useMemo(() => (id, type, action) => {
        dispatch({
            type: action,
            payload: {id, type}
        });
    }, [dispatch]);

    return {
        openDiagramView: (id, type) => handleDiagramView(id, type, 'OPEN_DIAGRAM'),
        minimizeDiagramView: (id, type) => handleDiagramView(id, type, 'MINIMIZE_DIAGRAM'),
        togglePinDiagramView: (id, type) => handleDiagramView(id, type, 'TOGGLE_PIN_DIAGRAM'),
        closeDiagramView: (id, type) => handleDiagramView(id, type, 'CLOSE_DIAGRAM'),
        closeDiagramViews: (idsToRemove) => handleDiagramView(idsToRemove, null, 'CLOSE_DIAGRAMS'),
    };
};
```

Reply:
Maybe in another TS, this would complexify the current refactoring too much.